### ISSUE TITLE
Unity builds for Windows with il2cpp support  

### ISSUE THREAD

**Issue by denisbut1212** [game-ci/unity-builder#82]
Is it possible to run builder action with [this](https://hub.docker.com/layers/gableroux/unity3d/2019.3.7f1-windows-99-try-il2cpp/images/sha256-174761b9d614e24c3ea50ead46527d03f32ac6b82107d0ded2ef0f400075306d?context=explore) docker image?

**Comment by GabLeRoux** [comment]
## Linked docker image details

FYI here are some details about that specific image:  

```bash
skopeo --override-os linux inspect docker://gableroux/unity3d:2019.3.7f1-windows-99-try-il2cpp | jq '.Labels'
```

```json
{
  "build-date": "2020-03-30T18:48:45+0000",
  "com.gableroux.unity3d.build": "f1",
  "com.gableroux.unity3d.download_url": "https://beta.unity3d.com/download/6437fd74d35d/UnitySetup-2019.3.7f1",
  "com.gableroux.unity3d.release_notes": "https://unity3d.com/unity/whats-new/2019.3.7f1",
  "com.gableroux.unity3d.release_url": "https://beta.unity3d.com/download/6437fd74d35d/public_download.html",
  "com.gableroux.unity3d.sha1": "4845b7bd6e3d1df813da2be5c1fa580d0c2855ad",
  "com.gableroux.unity3d.tag": "2019.3.7f1-windows",
  "com.gableroux.unity3d.version": "2019.3.7",
  "com.gitlab.ci.build": "https://gitlab.com/gableroux/unity3d/builds/491386513",
  "com.gitlab.ci.builder": "lebreton.gabriel@gmail.com",
  "com.gitlab.ci.pipeline": "https://gitlab.com/gableroux/unity3d/pipelines/131159034",
  "com.gitlab.ci.ref": "99-try-il2cpp",
  "vcs-url": "https://gitlab.com/gableroux/unity3d"
}
```

* image build log: https://gitlab.com/gableroux/unity3d/-/jobs/491386513
* related merge request: https://gitlab.com/gableroux/unity3d/-/merge_requests/99

Just to make sure this is not misleading, here is a quick explanation: there is `IL2CPP` and `windows` in the name of the image, but that doesn't mean it's an image that has support for creating windows builds with IL2CPP. `99-try-il2cpp` is a branch name and it was done to add Linux IL2CPP support.

## Can we create an IL2CPP windows build with gableroux/unity3d Docker images?

Short answer: No

gableroux/unity3d Docker images are based on linux (ubuntu:18.04 to be more precise [as shown here](https://gitlab.com/gableroux/unity3d/-/blob/1e3012f0/docker/ubuntu_prepare.Dockerfile#L1)). It's unity running in Docker on Linux so before investigation goes further, if we cannot create an IL2CPP windows build when running Unity on Linux, then it won't be possible to do it with any pipeline using gableroux/unity3d Docker images.

According to the following forum discussion:  
https://forum.unity.com/threads/cannot-make-a-build-in-linux-because-of-il2cpp.746192/

It is not possible to create windows builds with IL2CPP support using Unity running on Linux.

## Possible alternatives

_I have a meeting, but I'll edit this section in a few minutes_ ;)

**Comment by denisbut1212** [comment]
Possible alternatives?

**Comment by GabLeRoux** [comment]
That was a very long meeting ahahah. Alright, so here's the **Possible alternatives**. Note, they are overkill and may not work or be good solutions.

This is going to be a braindump from nights of trying things that leaded nowhere so far. Also note that I don't plan on supporting anyone here with what I'm about to share nor maintain the scripts I'm sharing here. This is not what I'm using right now. After spending so much time trying to get things working with windows, I just went the manual way. It is much easier even tho it pisses me off considering my CI knowledge and background ;)

## 1. Personal computer windows runner: cheap solution, hard to automate

I don't know about windows support and github actions (to be honest, I didn't try Github Actions a lot yet) so this may not be possible with github actions (probably a google search away from the answer), but I guess developers always find a way. 

So here's the thing: One can configure his build machine running on windows and install everything necessary for its unity project locally. This is the same problem as building for consoles and other targets that require windows to build. The idea is simply to not use the docker images I published or a unity installation running on Linux to build the project. Just use a windows machine that works.

There are many ways to execute scripts on a remote windows system. I have more experience with gitlab and can explain what I tried so far, but I did face many problems so the following may not be reliable at all.

I used to setup a gitlab runner which basically is a local machine at home with everything installed and is able to run powershell or bash and with some magic tricks, git bash. It is a machine connected to my gitlab organization as a personal runner that is allowed to pick CI jobs and execute them when asked for it.

### Registering a windows gitlab runner

https://github.com/GabLeRoux/unity3d-windows-build-scripts-drafts/blob/master/register-windows-gitlab-runner.ps1

So for my machine, I had two runners, one for running `powershell` and one for running `bash`. Now it says bash but it's actually WSL (See section bellow for more details on WSL).

### Magic trick to use git bash instead of WSL

https://gitlab.com/gitlab-org/gitlab-runner/issues/1703#note_43132602

Change default runner (`cmd.exe` as admin):

View current value

```powershell
reg query HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\gitlab-runner /v ImagePath
```

Tried this but it failed:

```powershell
reg add HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\gitlab-runner /f /v ImagePath /t REG_EXPAND_SZ /d "\"C:\Program Files\GitLab\gitlab-runner.exe\" run --working-directory \"C:\Program Files\GitLab\" --config \"C:\Program Files\GitLab\config.toml\" --service gitlab-runner --syslog"
```

So I just did it by hand. The thing about automation is if it takes 100x more time to automate it, just don't.

### Magic trick to use git bash instead of WSL (manual method)

win+r: `regedit`

`HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\gitlab-runner`

`ImagePath`
```
C:\Program Files\Git\bin\bash.exe -c "/c/gitlab/gitlab-runner.exe run --working-directory /c/gitlab --config /c/gitlab/config.toml --service gitlab-runner"
```

_original value:_  
```
"C:\Program Files\GitLab\gitlab-runner.exe" run --working-directory "C:\Program Files\GitLab" --config "C:\Program Files\GitLab\config.toml" --service gitlab-runner --syslog
```

```
.\gitlab-runner.exe stop
.\gitlab-runner.exe start
```

### Powershell executor

I first tried with Powershell and gitlab-ci, but always faced permission errors. Here's the script I managed to write:

https://github.com/GabLeRoux/unity3d-windows-build-scripts-drafts/blob/master/build-windows-powershell.ps1

I did try to use https://github.com/Microsoft/unitysetup.powershell but for some reason, I was never able to get the module to work when powershell script was invoked by the gitlab runner so I tried with `New-Object System.Diagnostics.Process`. I don't remember the result exactly, but it was such a mess that I tried the WSL way.

Executing this script locally with powershell did work on my system, but invoking it from the CI gave me permission errors and other troubles so I flipped some tables and tried something else.

### WSL (Windows Subsystem for Linux) Executor

Version 2 of WSL was out recently when I tried this. I managed to write some script that was very similar to what I did in here: https://gitlab.com/gableroux/unity3d-gitlab-ci-example/-/blob/896c8cae/ci/build.sh and then converted paths from WSL to Linux and back to Windows with `wslpath` command line. More details on how to use here:  
https://superuser.com/a/1317102/55267

I probably have that script somewhere in my project's git history, but it wasn't a success so I went the `git bash` way which I find easier to use.

### Git bash executor

See `Magic trick to use git bash instead of WSL` section above to get this working, otherwise, gitlab-runner on windows won't run in git bash by default.

This script is probably what managed to be the solution that was as close as possible to a working solution in my case:  
https://github.com/GabLeRoux/unity3d-windows-build-scripts-drafts/blob/master/build-windows-git-bash.sh

If you read through it, you'll see that I had to use a workaround to actually read the logs on windows because there is/was [a bug where unity invoked by command line on windows doesn't log to standard output](https://issuetracker.unity3d.com/issues/command-line-logfile-with-no-parameters-outputs-to-screen-on-os-x-but-not-on-windows) (cool cool cool, now this is very easy to use in a CI... (not)). It says it was fixed in `2019.1` (I did not try it again, got feature freeze in a 2018 version in my case). Anyway, someone seems to have reported stdout working in [the following comment](https://forum.unity.com/threads/redirecting-standard-output-using-the-logfile-parameter-when-in-batchmode.395339/#post-5573377)

## 2. Don't use a CI, simply remote control a build machine already setup

In my case, I simply remote control the build machine with Windows Remote Desktop App. Team viewer can do it too. `git push`, remote control, `git pull`, open project with unity and everything already configured, hit build. Optionally execute a script to upload your build somewhere. This is actually what I do and it's good enough for me. I wish I had a better solution ;)

## 3. Build AWS AMIs with pre-installed versions of Unity

This is an idea I had a while ago. It is technically possible to use something like [Packer](https://www.packer.io/) to build [AWS AMIs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIs.html) which are simply images for EC2 instances (servers) and then use some scripts or tools like https://github.com/Microsoft/unitysetup.powershell to install unity in these images and then publish them.

Thing is running Windows in the cloud is expensive. Maybe a bit cheaper with Azure, but again, for smaller teams or indie studios, the have a local machine we maintain by hand isn't such a bad thing as it just works.

With AWS, it would then be possible to spin [AWS EC2 spot instances](https://aws.amazon.com/ec2/spot/) (cheaper instances that can be reserved for few hours, handy for CI workload) and have them build the project with a solution like the ones I suggested above.

Now this sounds really great in theory, it's probably possible to do it, but there are so many unity version and so many components and some of them requires special NDAs sometimes to actually install them so these could not be published publicly anyway so efforts are probably too high for what we would gain here.

Still, this really sound like a nice weekend project and I will maybe try to write something about this some day :)

## 4. Train some AI so it can replace a human to build projects

No, I'm just trolling, but we already have self-driving cars, who knows?

## 5. Dockerize windows

I did not explore this at all. Maybe it's expensive to run this in the cloud too, I just don't know. Anyway, one could buy and read https://www.amazon.com/gp/product/1785281658/

There are actually [Windows base OS images on docker hub](https://hub.docker.com/_/microsoft-windows-base-os-images)

But the thing is this:

> Windows requires the host OS version to match the container OS version. 

So in other words, it's like the `3. Build AWS AMIs with pre-installed versions of Unity` where I suggest to use Packer to build AMIs, but instead, we would use Docker. Result would be very similar, but images would be on docker hub instead of AMIs AWS Registries.

## 6. UI Automation

Like `2. Don't use a CI, simply remote control a build machine already setup`, but script something that clicks on buttons for you. Please don't do this :joy:

---

That's it for the braindump. Maybe we'll get somewhere with that some day 👀 🤞

**Comment by RDeluxe** [comment]
Hey !

I'm sorry, but isn't this issue resolved in a way thanks to https://gitlab.com/gableroux/unity3d/-/issues/66 ?

**Comment by davidmfinol** [comment]
It is possible to get windows builds with il2cpp in GitHub Actions. 
With @SLIB53's help,  we got a workflow with a job like this working for me:
```
  buildForWindows:
    runs-on: windows-latest
    env:
      UNITY_USERNAME: ${{ secrets.UNITY_EMAIL }}
      UNITY_PASSWORD: ${{ secrets.UNITY_PASSWORD }}
      UNITY_SERIAL: ${{ secrets.UNITY_SERIAL }}
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v2
        with:
          lfs: true
      - name: Install Unity
        uses: crazy-max/ghaction-chocolatey@v1.2.2
        with:
          args: install --yes unity --version=2019.4.1 unity-il2cpp --version=2019.4.1 unity-win-il2cpp --version=2019.4.1 unity-metro --version=2019.4.1
      - name: Activate License
        run: Start-Process -FilePath "C:\Program Files\Unity\Editor\Unity.exe" -ArgumentList "-batchmode","–quit","-logfile activate.stdout.txt","-nographics","-username $env:UNITY_USERNAME","-password $env:UNITY_PASSWORD","-serial $env:UNITY_SERIAL" -NoNewWindow -Wait
      - name: Build StandaloneWindows
        run: Start-Process -FilePath "C:\Program Files\Unity\Editor\Unity.exe" -ArgumentList "-batchmode","–quit","-logfile build-windows.stdout.txt","-executeMethod Cgs.Editor.BuildCgs.BuildWindows" -NoNewWindow -Wait
      - name: Build StandaloneWindows64
        run: Start-Process -FilePath "C:\Program Files\Unity\Editor\Unity.exe" -ArgumentList "-batchmode","–quit","-logfile build-windows64.stdout.txt","-executeMethod Cgs.Editor.BuildCgs.BuildWindows64" -NoNewWindow -Wait
      - name: Build Uwp
        run: Start-Process -FilePath "C:\Program Files\Unity\Editor\Unity.exe" -ArgumentList "-batchmode","–quit","-logfile build-uwp.stdout.txt","-executeMethod Cgs.Editor.BuildCgs.BuildUwp" -NoNewWindow -Wait
      - name: Return License
        run: Start-Process -FilePath "C:\Program Files\Unity\Editor\Unity.exe" -ArgumentList "-batchmode","–quit","-logfile return.stdout.txt","-returnlicense" -NoNewWindow -Wait
      - name: Print Logs
        run: |
          cat activate.stdout.txt
          cat build-windows.stdout.txt
          cat build-windows64.stdout.txt
          cat build-uwp.stdout.txt
          cat return.stdout.txt
```

For reference, this is what the build methods look like in Cgs.Editor.BuildCgs.cs:
```
        public static void BuildWindows()
        {
            Build(BuildTarget.StandaloneWindows, Path.Combine(Path.Combine("builds", "windows"), "cgs.exe"));
        }
        public static void BuildWindows64()
        {
            Build(BuildTarget.StandaloneWindows64, Path.Combine(Path.Combine("builds", "windows64"), "cgs.exe"));
        }
        public static void BuildUwp()
        {
            EditorUserBuildSettings.wsaUWPBuildType = WSAUWPBuildType.XAML;
            Build(BuildTarget.WSAPlayer, Path.Combine("builds", "uwp"));
        }
        private static void Build(BuildTarget buildTarget, string filePath)
        {
            string[] scenes = EditorBuildSettings.scenes.Where(scene => scene.enabled).Select(s => s.path).ToArray();
            var buildPlayerOptions = new BuildPlayerOptions
            {
                scenes = scenes,
                locationPathName = filePath,
                target = buildTarget,
            };
            BuildReport buildReport = BuildPipeline.BuildPlayer(buildPlayerOptions);
            BuildSummary summary = buildReport.summary;
            ReportSummary(buildReport.summary);
            ExitWithResult(summary.result);
        }
```

I don't know if this is useful for Unity Actions, but I thought it'd be good to share.

**Comment by webbertakken** [comment]
Thanks a bunch for sharing a working workflow for windows il2cpp @davidmfinol.

Definitely very useful to have references.

The biggest challenge for running that in unity-actions is that it's based (and will stay based) on docker images. The problem is that the docker image will have to be windows based and maybe even have to run on a windows machine.

**Comment by davidmfinol** [comment]
Some extra resources on potentially creating a Windows Docker image:
https://azure.github.io/AppService/2020/06/09/App-Service-Continuous-Deployment-for-Windows-Containers-with-GitHub-Actions.html
https://docs.github.com/en/actions/reference/specifications-for-github-hosted-runners
https://hub.docker.com/_/microsoft-windows-servercore?tab=description

Since `runs: windows-latest` uses a Windows Server 2019 image, and Windows Docker images are required to use the same OS as the Windows host, we would likely need to base our Dockerfile `FROM mcr.microsoft.com/windows/servercore:ltsc2019`.

**Comment by webbertakken** [comment]
A few more pieces of information on Unity's "[official umbrella issue](https://issuetracker.unity3d.com/issues/cannot-activate-license-within-a-docker-container)".

What caught my attention:

Cause:

> we have discovered that the reason that the license activation fails from within this docker instance is that our implementation of `GetInternetReachability` seems to fail within the docker container.

Introduced in:

> 2019.3+

Fixed in:

> 2019.4.latest
> 2020.2.latest

Solution:

> Re-enabling netprofm is insufficient, as it depends on another service (nlasvc), which must also be re-enabled.
> Furthermore, another critical service is also required, by licensing and perhaps elsewhere: WMI (wmiApSrv).

Thus: enable `netprofm`, `nlasvc` and `wmiApSrv` in the Windows container. (and run it on top of a Windows machine)

**Comment by chrishultin** [comment]
To enable the services via PowerShell inside of a Windows-based container:
`Get-Service -Name "wmiapsrv" | Set-Service -Status running -StartupType automatic
Get-Service -Name "nlasvc" | Set-Service -Status running -StartupType automatic
Get-Service -Name "netprofm" | Set-Service -Status running -StartupType automatic
`

**Comment by chrishultin** [comment]
Here's an example Dockerfile that installs Unity silently to the default directory:

```Dockerfile
FROM mcr.microsoft.com/windows:1909

RUN powershell.exe -command (New-Object System.Net.WebClient).DownloadFile('https://download.unity3d.com/download_unity/9957aee8edc2/Windows64EditorInstaller/UnitySetup64-2020.1.17f1.exe', '.\UnitySetup64-2020.1.17f1.exe')
RUN powershell.exe -command (New-Object System.Net.WebClient).DownloadFile('https://download.unity3d.com/download_unity/9957aee8edc2/TargetSupportInstaller/UnitySetup-Universal-Windows-Platform-Support-for-Editor-2020.1.17f1.exe', '.\UnitySetup-Universal-Windows-Platform-Support-for-Editor-2020.1.17f1.exe')

RUN UnitySetup64-2020.1.17f1.exe /S
RUN UnitySetup-Universal-Windows-Platform-Support-for-Editor-2020.1.17f1.exe /S
```

The final path to the Unity executable is `C:\Program Files\Unity\Editor\Unity.exe`

**Comment by davidmfinol** [comment]
Closing this issue, as windows builds with IL2CPP are now possible.
Related issue for documentation: https://github.com/game-ci/documentation/issues/239
